### PR TITLE
coap-gateway: disconnection tls connection for expiration ceritificate

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

**Review by Danielius1922 — APPROVED** [review]

File: charts/plgd-hub/templates/coap-gateway/config.yaml [hunk 1]
```diff
@@ -39,6 +39,7 @@ data:
           blockSize: {{ .apis.coap.blockwiseTransfer.blockSize | quote }}
         tls:
           enabled: {{ .apis.coap.tls.enabled }}
+          disconnectForCertificateExpiration: {{ .apis.coap.tls.disconnectForCertificateExpiration }}
```
Comment (ondrejtomcik) [inline charts/plgd-hub/templates/coap-gateway/config.yaml]: `disconnectOnExpiredCertificate`

File: coap-gateway/service/auth.go [hunk 2]
```diff
@@ -69,9 +68,9 @@ func (s *Service) VerifyDeviceID(tlsDeviceID string, claim pkgJwt.Claims) error
 	return nil
 }
 
-func verifyChain(chain []*x509.Certificate, capool *x509.CertPool) (string, error) {
+func verifyChain(chain []*x509.Certificate, capool *x509.CertPool) error {
 	if len(chain) == 0 {
-		return "", fmt.Errorf("empty chain")
+		return fmt.Errorf("empty chain")
```
Comment (ondrejtomcik) [inline coap-gateway/service/auth.go]: certificate chain is empty

File: coap-gateway/service/auth.go [hunk 3]
```diff
@@ -99,37 +98,37 @@ func verifyChain(chain []*x509.Certificate, capool *x509.CertPool) (string, erro
 		}
 	}
 	if !ekuHasClient {
-		return "", fmt.Errorf("not contains ExtKeyUsageClientAuth")
+		return fmt.Errorf("not contains ExtKeyUsageClientAuth")
```
Comment (ondrejtomcik) [inline coap-gateway/service/auth.go]: the extended key usage field in the device certificate does not contain client authentication

File: coap-gateway/service/auth.go [hunk 4]
```diff
@@ -99,37 +98,37 @@ func verifyChain(chain []*x509.Certificate, capool *x509.CertPool) (string, erro
...
 	}
 	if !ekuHasServer {
-		return "", fmt.Errorf("not contains ExtKeyUsageServerAuth")
+		return fmt.Errorf("not contains ExtKeyUsageServerAuth")
```
Comment (ondrejtomcik) [inline coap-gateway/service/auth.go]: the extended key usage field in the device certificate does not contain server authentication

File: coap-gateway/service/auth.go [hunk 5]
```diff
@@ -99,37 +98,37 @@ func verifyChain(chain []*x509.Certificate, capool *x509.CertPool) (string, erro
...
 	}
-	return coap.GetDeviceIDFromIdentityCertificate(certificate)
+	_, err = coap.GetDeviceIDFromIdentityCertificate(certificate)
+	if err != nil {
+		return fmt.Errorf("cannot get deviceID from common name: %w", err)
```
Comment (ondrejtomcik) [inline coap-gateway/service/auth.go]: the device ID not part of the certificate's common name

File: coap-gateway/service/client.go [hunk 6]
```diff
@@ -87,6 +87,7 @@ type Client struct {
 	server      *Service
 	coapConn    *tcp.ClientConn
 	tlsDeviceID string
+	tlsValidTo  time.Time
```
Comment (ondrejtomcik) [inline coap-gateway/service/client.go]: validUntil

**Review by ondrejtomcik — CHANGES_REQUESTED** [review]

File: coap-gateway/service/client.go (hunk 6 above)
Comment (jkralik) [reply coap-gateway/service/client.go]: Fixed

File: coap-gateway/service/auth.go (hunk 4 above)
Comment (jkralik) [reply coap-gateway/service/auth.go]: Fixed

File: coap-gateway/service/auth.go (hunk 5 above)
Comment (jkralik) [reply coap-gateway/service/auth.go]: Fixed

File: coap-gateway/service/auth.go (hunk 3 above)
Comment (jkralik) [reply coap-gateway/service/auth.go]: Fixed

File: coap-gateway/service/auth.go (hunk 2 above)
Comment (jkralik) [reply coap-gateway/service/auth.go]: Fixed

File: charts/plgd-hub/templates/coap-gateway/config.yaml (hunk 1 above)
Comment (jkralik) [reply charts/plgd-hub/templates/coap-gateway/config.yaml]: Fixed

**Review by ondrejtomcik — APPROVED** [review]

**Pull request merged**